question-servers/freeform.js

```diff
@@ -1093,17 +1057,36 @@ module.exports = {
                 options: _.get(variant, 'options', {}),
                 filename: filename,
             };
-            workers.getPythonCaller((err, pc) => {
-                if (ERR(err, callback)) return;
-                module.exports.processQuestion('file', pc, data, context, (err, courseIssues, _data, _html, fileData) => {
-                    // don't immediately error here; we have to return the pythonCaller
-                    workers.returnPythonCaller(pc, (pcErr) => {
-                        if (ERR(pcErr, callback)) return;
+
+            module.exports.getCachedDataOrCompute(
+                course,
+                data,
+                context,
+                (callback) => {
+                    // function to compute the file data and return the cachedData
+                    workers.getPythonCaller((err, pc) => {
                         if (ERR(err, callback)) return;
-                        callback(null, courseIssues, fileData);
+                        module.exports.processQuestion('file', pc, data, context, (err, courseIssues, _data, _html, fileData) => {
+                            // don't immediately error here; we have to return the pythonCaller
+                            workers.returnPythonCaller(pc, (pcErr) => {
+                                if (ERR(pcErr, callback)) return;
+                                if (ERR(err, callback)) return;
+                                const fileDataBase64 = fileData.toString('base64');
+                                const cachedData = {courseIssues, fileDataBase64};
+                                callback(null, cachedData);
+                            });
+                        });
                     });
-                });
-            });
+                },
+                (cachedData, _cacheHit) => {
+                    // function to process the cachedData, whether we
+                    // just rendered it or whether it came from cache
+                    const {courseIssues, fileDataBase64} = cachedData;
+                    const fileData = Buffer.from(fileDataBase64, 'base64');
```

Comment:
I haven't played around with the caching code so maybe this is an obvious question, but is it necessary to convert to/from base64 when storing the file data?

Reply:
My initial local testing with the memory cache worked fine leaving the `fileData` as a buffer, but then when I tried it on `pl-dev` (using redis) it didn't work. Changing to base64 fixed it for redis, and it still worked with the memory cache as well. I didn't investigate to figure out why redis had a problem with binary data.

Reply:
Strange, but I suppose if it works... 🤷

Reply:
Redis strings are supposedly binary safe, so maybe the issue arose in some other part of the toolchain.